### ISSUE TITLE
Code Lab Feedback

### ISSUE THREAD

**Issue by zacharyweston** [zacharyweston/codelab-jquery-intro#1]
Hey @chrisvfritz, can you take a look at this? I have been having no luck trying to get the if statement to work, and I cannot figure it out

**Comment by chrisvfritz** [comment]
Happy to. :smiley:

I found the first problem by [opening up Chrome's JavaScript console](https://developers.google.com/web/tools/chrome-devtools/debug/console/console-ui?hl=en#opening-the-console) and I saw this error: 

```
Uncaught ReferenceError: $ is not defined                                     (index):44
```

That `44` at the end is referring to the line number the error is occurring on in your `index.html` file. Which is this one:

``` js
var submission1 = $('#input1').text();
```

The problem is that when it gets to `$`, it has no idea what it is. If you saw this, you were probably baffled thinking, "But that's jQuery! And I included jQuery in my head, didn't I?"

You did, but as you've now learned from experience, strange things can happen in browsers when you have _some_ scripts in the head and _some_ scripts in the body. So I moved your jQuery script to your body, right before your own script tag:

``` html
<script src="http://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script>
  var submission1 = $('#input1').text();
  $('#submit1').on('click', function() {
    if ($('#input1').text() == "x") {
      $('#square1').css('background-color', '#b0e0e6');
    }
  });
</script>
```

After that, I didn't see any errors. But when I tried to put an `X` into `#input1`, nothing happened. I checked the JavaScript console and there were no errors there.

So then I copied some code from line 44 and pasted it into the JavaScript console:

``` js
$('#input1').text()
```

And I got back `""` - an empty string. Even though I _clearly_ see an `X` in the input. At this point, I would google something like ["jquery input text returns empty string"](https://www.google.com/search?q=jquery+input+text+returns+empty+string). These are the results I got back:

![jQuery troubleshooting Google results](https://www.dropbox.com/s/0alfle62qc9sp5f/Screenshot%202016-01-21%2019.23.33.png?dl=1)

The second result looks somewhat similar to what we're experiencing. Diving in, we can see someone complaining of a very similar problem and the accepted answer is:

![jQuery troubleshooting StackOverflow answer](https://www.dropbox.com/s/yd7fejlonkxizqk/Screenshot%202016-01-21%2019.26.57.png?dl=1)

We're using a an `input` rather than a `textarea`, but let's try using `.val()` in the JavaScript console to test it out:

``` js
$('#input1').val();
```

And we got back `"X"` just like we wanted! I hope that gets you unstuck. Please don't hesitate to let me know if some part of that exploration didn't make sense to you or if you run across another roadblock. This lesson can be a tricky one!

**Comment by zacharyweston** [comment]
Thanks for the help, that was a huge sticking point, and your explanation
was super helpful!

**Comment by zacharyweston** [comment]
New code has been pushed, and the URL is http://bright-gold.surge.sh/

**Comment by chrisvfritz** [comment]
This works perfectly! Also nice job taking advantage of the inherent nature of a webpage - your solution to reload the page by just refreshing is very elegant! There are a couple ways this could be made even simpler though and it'll be a good chance to learn a little more about jQuery, so I'm going to push us a little further.

## No need to reset input values when refreshing the page

When the page is refreshed, all the input values are already emptied, because we're not saving their value in the URL parameters. So this:

``` js
$('#resetButton').on('click', function() {
  $('#input1').val('');
  $('#input2').val('');
  $('#input3').val('');
  $('#input4').val('');
  $('#input5').val('');
  $('#input6').val('');
  $('#input7').val('');
  $('#input8').val('');
  $('#input9').val('');
  location.reload();
});
```

can actually be shortened to:

``` js
$('#resetButton').on('click', function() {
  location.reload();
});
```

Refreshing the page will do the job on its own! :smiley: 

## Avoiding repeated logic in jQuery

When you find yourself copy and pasting the same code over and over again and just changing a number, that's often a good sign that your code can be made more generic. Starting with the HTML, each table cell currently looks something like this:

``` html
<td id="square1">
  <input type="text" id="input1" maxlength="1" size="1">
  <input type="submit" value="Submit" id="submit1">
</td>
```

Since all of our cells should act the same, let's make this more generic by changing our `id`s to `class`es (since we want many of them) and removing the numbers.

``` html
<td class="square">
  <input type="text" class="input" maxlength="1" size="1">
  <input type="submit" value="Submit" class="submit">
</td>
```

And this can actually be made a little simpler. There's a CSS selector you may not have encountered yet to target items by an attribute. If we want to find all `input`s of `type` `text`, we can use `input[type="text"]`. To target all `input`s of `type` `submit`, we'd use `input[type="submit"]`. So let's remove those classes, shortening this to just:

``` html
<td class="square">
  <input type="text" maxlength="1" size="1">
  <input type="submit" value="Submit">
</td>
```

Alright, now for the JavaScript! If we were describing the behavior we want every table cell to have, we might say, _When the submit button inside a table cell is clicked, we should check the value of the text input and set the background color of the cell accordingly._ 

So first, let's capture when the submit button inside a `.square` is clicked:

``` js
$('.square > input[type="submit"]').on('click', function() {
  console.log('Clickity click!');
});
```

Pressing those buttons and checking the JavaScript console, I am indeed seeing that message, so it seems to be working.

Now I want to get the value of the `text` input. But here's the problem. I can't do something similar and use `$('.square > input[type="text"]')`, because that would get _every_ text input inside the `.square`. I just want the one right next to the button that was actually clicked.

For cases just like this, there's a special variable inside jQuery event handlers called `this`, which represents the element that was clicked on. We can turn that element into a jQuery object with `$(this)`. And once we do that, we can explore the HTML around it, for example using the `.siblings()` method to find a sibling (adjacent) element that happens to be our text input.

Let's try to do that below, retrieve the value from that element, then log it to see if we're getting what we expect:

``` js
$('.square > input[type="submit"]').on('click', function() {
  var $submitButton = $(this);
  var submission = $submitButton.siblings('input[type="text"]').val();
  console.log(submission);
});
```

That works! Now we just need one more element - the `.square` container - so that we can change its background. To get the containing element, there's another method called `.parent()`, which doesn't need any arguments.

I add this below, along with another `console.log()` just to make sure it's working.

``` js
$('.square > input[type="submit"]').on('click', function() {
  var $submitButton = $(this)
  var submission = $submitButton.siblings('input[type="text"]').val();
  var $squareCell = $submitButton.parent();
  console.log($squareCell);
});
```

Now to complete that whole `click` handler:

``` js
$('.square > input[type="submit"]').on('click', function() {
  var $submitButton = $(this)
  var submission = $submitButton.siblings('input[type="text"]').val();
  var $squareCell = $submitButton.parent();
  if (submission === 'X' || submission === 'x') {
    $squareCell.css('background-color', '#b0e0e6');
  } else if (submission === 'O' || submission === 'o') {
    $squareCell.css('background-color', '#C9AEE5');
  }
});
```

That's looking really good and seems to work, but there's just one more optimization we can make. Let's capitalize the submission with `.toUpperCase()` so that we don't have to check for both upper and lower case versions of `X` and `O`.

``` js
$('.square > input[type="submit"]').on('click', function() {
  var $submitButton = $(this)
  var submission = $submitButton.siblings('input[type="text"]').val().toUpperCase();
  var $squareCell = $submitButton.parent();
  if (submission === 'X') {
    $squareCell.css('background-color', '#b0e0e6');
  } else if (submission === 'O') {
    $squareCell.css('background-color', '#C9AEE5');
  }
});
```

So now our _entire_ program is just:

``` js
$('#resetButton').on('click', function() {
  location.reload();
});

$('.square > input[type="submit"]').on('click', function() {
  var $submitButton = $(this)
  var submission = $submitButton.siblings('input[type="text"]').val().toUpperCase();
  var $squareCell = $submitButton.parent();
  if (submission === 'X') {
    $squareCell.css('background-color', '#b0e0e6');
  } else if (submission === 'O') {
    $squareCell.css('background-color', '#C9AEE5');
  }
});
```

That's 94 lines reduced to just 10 lines of much simpler code! :tada: As you implement that, let me know if there's any part of the process you didn't understand and I'll be happy to clarify. :smiley:

**Comment by zacharyweston** [comment]
Makes sense, thanks for the instructions! New code has been pushed, and the new URL is http://wrong-bag.surge.sh/

**Comment by chrisvfritz** [comment]
Looks great! :tada: :balloon: :shipit: Regarding our conversation about ideal ways to check for win conditions, here are some good threads on the topic in case you want to dive deeper:
- [the math perspective](http://math.stackexchange.com/questions/467757/determine-the-winner-of-a-tic-tac-toe-board-with-a-single-matrix-expression)
- [the comp sci perspective](http://stackoverflow.com/questions/1056316/algorithm-for-determining-tic-tac-toe-game-over)

If I were programming a real tic-tac-toe app though, I would just Google for a little bit and then steal the solution that someone else already spent way too much time on. :smiley: In the cases where you can't find one though, it's almost always a better idea to program a clumsy solution that works (and that you'll still understand a month from now), than to learn advanced math or computer science concepts to arrive at a super clever solution that you won't understand a month from now.
